### ISSUE TITLE
public access to the foreign mapper class or classname in MapperRelationship

### ISSUE THREAD

**Issue by far-blue** [atlasphp/Atlas.Mapper#21]
Hello,

I'm trying to remove some repetitive, boilerplate code in our app models and one of the areas we see a lot of repeated code is in handling the creation of new records with related fields. When you create a new record none of the related fields will be setup - they are all null - so when we come to set values on them we first need to create the appropriate Record or RecordSet class. While I don't intend to blanket-create these empty classes on record creation I would like to create a simple function that will handle it.

At the moment there is no way to retrieve the foreign mapper class name so any util code would still need to accept the class of the foreign mapper as a parameter. If either the MapperRelationships::getForeignMapper() function was public or there was a public MapperRelationships::getForeignMapperClass() method similar to the Record::getMapperClass() method then I could remove a lot of hard-coded class names and simplify our code by being able to look up the mapper class based on the related field name.

The only way to achieve something similar at the moment is via the stitchIntoRecords() method but this makes unneeded db calls.

Would anyone object to such as change?

**Comment by froschdesign** [comment]
> If either the MapperRelationships::getForeignMapper() function was public…

This method existed in version 1 and 2:

```php
 /**
  *
  * Returns the foreign Mapper instance.
  *
  * @return MapperInterface
  *
  */
 public function getForeignMapper();
```

https://github.com/atlasphp/Atlas.Orm/blob/bf689041350ef81936f379db8f7699b806d59320/src/Relationship/RelationshipInterface.php#L50-L57
https://github.com/atlasphp/Atlas.Orm/blob/5edf8a93f7e8ff93cd0a46953d6ec703d03b3549/src/Relationship/RelationshipInterface.php#L62-L69

But was removed with version 3. [edit]

**Comment by froschdesign** [comment]
> I'm trying to remove some repetitive, boilerplate code in our app models and one of the areas we see a lot of repeated code is in handling the creation of new records with related fields.

Something like:

```php
$threadRecord->set([
    'topic'  => '…',
    'author' => [
        'name'  => '…'
        'email' => '…',
    ],
]);

```

**Comment by far-blue** [comment]
There's still support for providing the data for fields of a Record at the point of creation and it will go ahead and set everything up. We have written our domain model code to create empty Records and then use setters, however, because we treat create and update as the same. Creating an empty Record then "updating" it just like one fetched from the db (mostly) simplifies our domain level code.

Unfortunately, support for just providing the field names and not values (like when you fetch from the db) isn't there either.

**Comment by froschdesign** [comment]
Can you provide a code sample to illustrate your feature request? Or create a pull request with the desired changes.
Because at the moment, I'm afraid I don't know what exactly you want.

**Comment by far-blue** [comment]
I've raised PR #22 :)

**Comment by froschdesign** [comment]
I use this _unbeautiful / ugly_ trait:

```php
use Atlas\Mapper\Mapper;
use Atlas\Mapper\Relationship\RegularRelationship;
use ReflectionClass;

trait GetForeignerMapperTrait
{
    private function getForeignerMapper(
        RegularRelationship $relationship
    ): Mapper {
        $reflectionClass       = new ReflectionClass($relationship);
        $foreignerMapperMethod = $reflectionClass->getMethod(
            'getForeignMapper'
        );
        $foreignerMapperMethod->setAccessible(true);

        /** @var Mapper $foreignerMapper */
        return $foreignerMapperMethod->invoke($relationship);
    }
}
```

…but I hope to remove this **crap**!

@pmjones 
What do you think? Can we simplify the access / handling? And would that be in your interest?
Maybe you already have a better idea for version 4 (Dymaxion).

**Comment by pmjones** [comment]
@froschdesign @far-blue Are we satisfied that #22 resolves this issue?

**Comment by froschdesign** [comment]
@pmjones 
Works for me! 👍🏻

**Comment by far-blue** [comment]
As the author of the PR it's unsurprising I'm going to say yes ;)

**Comment by pmjones** [comment]
Fixed by #22
